Tiered storage

npawar opened the pull request:
## Description
Issue: https://github.com/apache/incubator-pinot/issues/5553
Tiered storage support in Pinot - Phase 1.
This phase supports default tag based instance assignments only, which are not persisted in zk. 

Phase 2 will introduce instanceAssignmentConfig for tiers, which will allow us to support replica groups for tiers and also let us persist the InstancePartitions

Example:
This example show how to configure segments older than 15 days move to `tier_b_OFFLINE` and segments older than 7 days move to `tier_a_OFFLINE`
```
{
  "tableName": "myTable",
  "tableType": ...,
  ...
  "tierConfigs": [{
    "name": "tierA",
    "segmentSelectorType": "time",
    "segmentAge": "7d",
    "storageType": "pinot_server",
    "serverTag": "tier_a_OFFLINE"
  }, {
    "name": "tierB",
    "segmentSelectorType": "TIME",
    "segmentAge": "15d",
    "storageType": "PINOT_SERVER",
    "serverTag": "tier_b_OFFLINE"
  }] 
}
```

## Release Notes
Tiered storage phase 1 - default tag based instance assignment for tiers

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -161,23 +185,80 @@ private void checkReplication(InstancePartitions instancePartitions) {
+   * Rebalances segments in the current assignment using the instancePartitions and returns new assignment
+   */
+  private Map<String, Map<String, String>> reassignSegments(String instancePartitionType,
+      Map<String, Map<String, String>> currentSegmentAssignment, InstancePartitions instancePartitions,
```
(nit) `currentAssignment` for concise?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/tier/TieredStorageRelocator.java:
```diff
@@ -0,0 +1,101 @@
+ * Periodic task to run rebalancer in background to relocate segments to storage tiers
+ * TODO: we could potentially get rid of tagOverrideConfig and rely on this relocator for moving COMPLETED segments
+ */
+public class TieredStorageRelocator extends ControllerPeriodicTask<Void> {
```
Please keep only one segment relocator. Currently there are 2 relocators: `TieredStorageRelocator` and `RealtimeSegmentRelocator`. They can rebalance the same table at the same time which could cause problem. Recommend replacing `RealtimeSegmentRelocator` with `SegmentRelocator` which handles both completed segment relocation and tier storage relocation.

npawar replied:
Removed TieredStorageRelocator. Rebranded RealtimeSegmentRelocator as `SegmentRelocator`.

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TierConfig.java:
```diff
@@ -0,0 +1,80 @@
+      @JsonProperty("segmentAge") @Nullable String segmentAge,
+      @JsonProperty(value = "storageType", required = true) String storageType,
+      @JsonProperty("serverTag") @Nullable String serverTag) {
+    _name = name;
```
`Preconditions.checkArgument()` on all non-null arguments to prevent bad config

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:
```diff
@@ -136,9 +143,55 @@ private static void validateIngestionConfig(@Nullable IngestionConfig ingestionC
+    for (TierConfig tierConfig : tierConfigList) {
+      String tierName = tierConfig.getName();
+      Preconditions.checkState(!tierName.isEmpty());
+      Preconditions.checkState(!tierNames.contains(tierName), "Tier name: %s already exists in tier configs", tierName);
+      tierNames.add(tierName);
```
(nit)
```suggestion
      Preconditions.checkState(tierNames.add(tierName), "Tier name: %s already exists in tier configs", tierName);
```

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -161,23 +185,80 @@ private void checkReplication(InstancePartitions instancePartitions) {
         config.getBoolean(RebalanceConfigConstants.BOOTSTRAP, RebalanceConfigConstants.DEFAULT_BOOTSTRAP);
+
+    Map<String, Map<String, String>> subsetAssignment = currentAssignment;
+    // rebalance tiers first
```
(nit) Capitalize the first letter for convention

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -161,23 +185,80 @@ private void checkReplication(InstancePartitions instancePartitions) {
     boolean bootstrap =
         config.getBoolean(RebalanceConfigConstants.BOOTSTRAP, RebalanceConfigConstants.DEFAULT_BOOTSTRAP);
+
+    Map<String, Map<String, String>> subsetAssignment = currentAssignment;
```
Rename to `nonTierAssignment` for clarity?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -161,23 +185,80 @@ private void checkReplication(InstancePartitions instancePartitions) {
+                tierName, _offlineTableName);
+        checkReplication(tierInstancePartitions);
+
+        LOGGER.info("Rebalancing tier: {} for table: {} with instance partitions: {}", tierName, _offlineTableName,
```
(nit) Including `bootstrap` in the log?

npawar commented on pinot-common/src/main/java/org/apache/pinot/common/assignment/InstanceAssignmentConfigUtils.java:
```diff
@@ -50,6 +50,13 @@ public static boolean shouldRelocateCompletedSegments(TableConfig tableConfig) {
+  /**
+   * Returns whether relocation of segments to tiers has been enabled for this table
+   */
+  public static boolean shouldRelocateToTiers(TableConfig tableConfig) {
```
Done

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -82,6 +91,10 @@ public void init(HelixManager helixManager, TableConfig tableConfig) {
         tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
     _partitionColumn = replicaGroupStrategyConfig != null ? replicaGroupStrategyConfig.getPartitionColumn() : null;
 
+    if (InstanceAssignmentConfigUtils.shouldRelocateToTiers(tableConfig)) {
```
Done. Added `sortedTiers` list to the rebalance method

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -91,6 +104,17 @@ public void init(HelixManager helixManager, TableConfig tableConfig) {
+   */
+  @VisibleForTesting
+  protected List<Tier> getSortedTiersForPinotServerStorage(List<TierConfig> tierConfigList) {
+    return tierConfigList.stream().filter(t -> TierFactory.PINOT_SERVER_STORAGE_TYPE.equals(t.getStorageType()))
```
Moved to TierConfigUtils, and removed lambdas

npawar commented:
> Well done.
> Most comments are minor (comments in `OfflineSegmentAssignment` also applies to `RealtimeSegmentAssignment`).
> Please merge the 2 relocators into one.

Thanks for the review. I've merged the 2 relocators into `SegmentRelocator`. I introduced controller confs for `controller.segment.relocator.frequency`, and deprecated the confs for `.controller.realtime.segment.relocator.frequency`. For some releases, we'll carry both around.

Jackie-Jiang commented on pinot-common/pom.xml:
```diff
@@ -182,6 +182,10 @@
       <groupId>org.apache.commons</groupId>
       <artifactId>commons-compress</artifactId>
     </dependency>
+    <dependency>
```
This is already included in the `pinot-spi`, no need to include again (you might need to rebase to the current master)

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/TierConfigUtils.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Util methods for TierConfig
+ */
+public class TierConfigUtils {
```
Add a private constructor

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java:
```diff
@@ -688,7 +684,7 @@ public static ControllerStarter startDefault(File webappPath) {
     conf.setBrokerResourceValidationFrequencyInSeconds(3600);
     conf.setStatusCheckerFrequencyInSeconds(5 * 60);
-    conf.setRealtimeSegmentRelocatorFrequency("1h");
+    conf.setSegmentRelocatorFrequencyInSeconds(60*60);
```
(nit) reformat (or just put 3600)

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java:
```diff
@@ -460,6 +459,30 @@ public void setStatusCheckerWaitForPushTimeInSeconds(int statusCheckerWaitForPus
+   * RealtimeSegmentRelocator has been rebranded to SegmentRelocator.
+   * Check for SEGMENT_RELOCATOR_FREQUENCY_IN_SECONDS property, if not found, return REALTIME_SEGMENT_RELOCATOR_FREQUENCY
+   */
+  public int getSegmentRelocatorFrequencyInSeconds() {
```
Well done

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/relocation/SegmentRelocator.java:
```diff
@@ -40,41 +41,51 @@
       ExecutorService executorService) {
-    super("RealtimeSegmentRelocator", getRunFrequencySeconds(config.getRealtimeSegmentRelocatorFrequency()),
-        config.getRealtimeSegmentRelocationInitialDelayInSeconds(), pinotHelixResourceManager, leadControllerManager,
+    super(SegmentRelocator.class.getSimpleName(), 60,
+        30, pinotHelixResourceManager, leadControllerManager,
```
**CRITICAL**
```suggestion
    super(SegmentRelocator.class.getSimpleName(), config.getSegmentRelocatorFrequencyInSeconds(),
        config.getSegmentRelocatorInitialDelayInSeconds(), pinotHelixResourceManager, leadControllerManager,
        controllerMetrics);
```

npawar replied:
removed!

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/tier/TierFactory.java:
```diff
@@ -0,0 +1,83 @@
+   * Tier with an older age bucket in {@link TimeBasedTierSegmentSelector} should appear before a younger age bucket, in sort order
+   * TODO: As we add more types, this logic needs to be upgraded
+   */
+  public static Comparator<Tier> getTierComparator() {
```
Move this into `TierConfigUtils`?

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/TierConfigUtils.java:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Gets tiers for given storage type from provided list of TierConfig
+   */
+  public static List<Tier> getTiersForStorageType(List<TierConfig> tierConfigList, String storageType,
```
Change this to `getSortedTiersForStorageType` and sort tiers here

npawar replied:
sure, done

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -161,19 +172,78 @@ private void checkReplication(InstancePartitions instancePartitions) {
+    Map<String, Map<String, String>> nonTierAssignment = currentAssignment;
+    // Rebalance tiers first
+    List<Map<String, Map<String, String>>> newTierAssignments = null;
+    if (TierConfigUtils.shouldRelocateToTiers(_tableConfig)) {
```
`shouldRelocateToTiers` logic is handled on the caller side, so here we can just check whether `sortedTiers` is not `null`. Same for realtime

Jackie-Jiang commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ControllerPeriodicTasksIntegrationTest.java:
```diff
@@ -111,8 +111,8 @@ public void setUp()
     properties.put(ControllerPeriodicTasksConf.STATUS_CHECKER_FREQUENCY_IN_SECONDS, PERIODIC_TASK_FREQUENCY_SECONDS);
-    properties.put(ControllerPeriodicTasksConf.REALTIME_SEGMENT_RELOCATION_INITIAL_DELAY_IN_SECONDS, PERIODIC_TASK_INITIAL_DELAY_SECONDS);
-    properties.put(ControllerPeriodicTasksConf.REALTIME_SEGMENT_RELOCATOR_FREQUENCY, PERIODIC_TASK_FREQUENCY);
+    properties.put(ControllerPeriodicTasksConf.DEPRECATED_REALTIME_SEGMENT_RELOCATION_INITIAL_DELAY_IN_SECONDS, PERIODIC_TASK_INITIAL_DELAY_SECONDS);
```
Should we use `SEGMENT_RELOCATION_INITIAL_DELAY_IN_SECONDS` and `SEGMENT_RELOCATOR_FREQUENCY_IN_SECONDS`?

npawar replied:
i kept the old ones, so that we have a test to check the old way, because people might be using the old properties. Probably after couple release can change it new?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignment.java:
```diff
@@ -60,9 +62,13 @@
   Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, String>> currentAssignment,
-      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap, Configuration config);
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap,
+      @Nullable Map<String, InstancePartitions> tierInstancePartitionsMap, @Nullable List<Tier> sortedTiers,
```
(nit) Slightly more readable if we put `sortedTiers` in front of `tierInstancePartitionsMap` as `tierInstancePartitionsMap` is describing the `sortedTiers`?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:
```diff
@@ -72,10 +81,12 @@
   private String _offlineTableName;
   private int _replication;
   private String _partitionColumn;
+  private TableConfig _tableConfig;
```
(OCD) move this to line 81

npawar replied:
even better, removed it because we no longer are checking `shouldRelocateToTiers(tableConfig)` in this file

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignment.java:
```diff
@@ -79,18 +88,33 @@
+   * Keeps only those which have "pinotServer" storage type.
+   */
+  @VisibleForTesting
+  protected List<Tier> getSortedTiersForPinotServerStorage(List<TierConfig> tierConfigList) {
```
Remove this method

npawar commented:
> LGTM otherwise. One **critical** comment in `SegmentRelocator`

Made one small change not part of the review. Converted the "segmentSelectorType" and "storageType" to enums. If you want to look again.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/tier/PinotServerTierStorage.java:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  @Override
+  public String getType() {
```
Should we return `TierStorageType` here?

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/tier/TierSegmentSelector.java:
```diff
@@ -0,0 +1,39 @@
+  /**
+   * The type of the segment selector (e.g. TIME)
+   */
+  String getType();
```
Return `TierSegmentSelectorType` here?

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TierConfig.java:
```diff
@@ -0,0 +1,84 @@
+  // TODO: only "serverTag" is supported currently. In next iteration, "InstanceAssignmentConfig _instanceAssignmentConfig" will be added here
+
+  public TierConfig(@JsonProperty(value = "name", required = true) String name,
+      @JsonProperty(value = "segmentSelectorType", required = true) String segmentSelectorType,
```
Enum instead of String?

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/TierConfigUtils.java:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Gets sorted list of tiers for given storage type from provided list of TierConfig
+   */
+  public static List<Tier> getSortedTiersForStorageType(List<TierConfig> tierConfigList, String storageType,
```
Use enum `storageType`

Jackie-Jiang commented:
> Made one small change not part of the review. Converted the "segmentSelectorType" and "storageType" to enums. If you want to look again.

Since we introduced the enum, try to use enum over string for these 2 fields

fx19880617 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignmentUtils.java:
```diff
@@ -326,4 +328,68 @@ static void rebalanceReplicaGroupBasedPartition(Map<String, Map<String, String>>
   }
+
+  /**
+   * Takes a segment assignment adn splits them up based on which tiers the segments are eligible for. Only considers ONLINE segments.
```
typo

npawar commented on pinot-common/src/main/java/org/apache/pinot/common/tier/TierSegmentSelector.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Interface for the segment selection strategy of a tier
+ */
+public interface TierSegmentSelector {
```
I thought I'll add that when the requirements demand it.. As of now, I didn't see it being requirwd

npawar commented:
> > Made one small change not part of the review. Converted the "segmentSelectorType" and "storageType" to enums. If you want to look again.
> 
> Since we introduced the enum, try to use enum over string for these 2 fields

The reason I kept it string is that if we want people to plug in their own strategies, they can do so without needing to add it to the enum

Jackie-Jiang commented:
> > > Made one small change not part of the review. Converted the "segmentSelectorType" and "storageType" to enums. If you want to look again.
> > 
> > 
> > Since we introduced the enum, try to use enum over string for these 2 fields
> 
> The reason I kept it string is that if we want people to plug in their own strategies, they can do so without needing to add it to the enum

In that case, IMO we should not introduce the enum

npawar commented:
> > > > Made one small change not part of the review. Converted the "segmentSelectorType" and "storageType" to enums. If you want to look again.
> > > 
> > > 
> > > Since we introduced the enum, try to use enum over string for these 2 fields
> > 
> > 
> > The reason I kept it string is that if we want people to plug in their own strategies, they can do so without needing to add it to the enum
> 
> In that case, IMO we should not introduce the enum

Hmm, agreed. Reverted enum change.

npawar commented on pinot-common/src/main/java/org/apache/pinot/common/tier/PinotServerTierStorage.java:
```diff
@@ -0,0 +1,46 @@
+ * Tier storage type which uses Pinot servers as storage
+ */
+public class PinotServerTierStorage implements TierStorage {
+  private final String _type = TierStorageType.PINOT_SERVER.toString();
```
Removed enums and kept just strings. When we eventually support users plugging in their own strategies, enum will restrict them from doing that

npawar commented on pinot-common/src/main/java/org/apache/pinot/common/tier/PinotServerTierStorage.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class PinotServerTierStorage implements TierStorage {
+  private final String _type = TierStorageType.PINOT_SERVER.toString();
+  private final String _tag;
```
Yes that's a good idea. I can add that for phase 2 (where I'll be handling advanced instance assignments for tiers)
